src/Components/Components/src/PersistentComponentState.cs

```diff
@@ -141,28 +133,41 @@ public void PersistState(string key, byte[] value)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            PersistState(key, JsonSerializer.SerializeToUtf8Bytes(instance, JsonSerializerOptionsProvider.Options));
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (_currentState.ContainsKey(key))
+            {
+                throw new ArgumentException($"There is already a persisted object under the same key '{key}'");
+            }
+
+            var writer = new PooledByteBufferWriter();
+            _currentState.Add(key, writer);
+            JsonSerializer.Serialize(new Utf8JsonWriter(writer), instance, JsonSerializerOptionsProvider.Options);
         }
 
         /// <summary>
         /// Tries to retrieve the persisted state as JSON with the given <paramref name="key"/> and deserializes it into an
         /// instance of type <typeparamref name="TValue"/>.
         /// When the key is present, the state is successfully returned via <paramref name="instance"/>
-        /// and removed from the <see cref="ComponentApplicationState"/>.
+        /// and removed from the <see cref="PersistentComponentState"/>.
         /// </summary>
         /// <param name="key">The key used to persist the instance.</param>
         /// <param name="instance">The persisted instance.</param>
         /// <returns><c>true</c> if the state was found; <c>false</c> otherwise.</returns>
-        public bool TryTakeAsJson<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(string key, [MaybeNullWhen(false)] out TValue? instance)
+        public bool TryTakeFromJson<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(string key, [MaybeNullWhen(false)] out TValue? instance)
```

Comment:
```suggestion
        public bool TryTakeFromJson<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(string key, [NotNullWhen(true)] out TValue? instance)
```

Reply:
Unfortunately this is not the case. `JsonSerializer.Deserialize` could return null